Tests/BackendIntegrationTests/VirtualCurrencyIntegrationTests.swift

```diff
@@ -0,0 +1,179 @@
+        let userID1 = "vc_user_\(UUID().uuidString)"
+        let userID2 = "vc_user_\(UUID().uuidString)"
+
+        _ = try await self.purchases.logIn(userID1)
+
+        try self.purchases.invalidateVirtualCurrenciesCache()
+        let virtualCurrenciesBeforePurchase = try await self.purchases.virtualCurrencies()
+        assertAllVirtualCurrenciesHaveZeroBalances(virtualCurrenciesBeforePurchase)
+
+        try await self.purchaseNonRenewingSubscriptionPackage()
+        self.verifyTransactionWasFinished()
+
+        try self.purchases.invalidateVirtualCurrenciesCache()
+
+        let virtualCurrenciesAfterPurchase = try await self.purchases.virtualCurrencies()
+
+        expect(virtualCurrenciesAfterPurchase["TEST"]?.balance).to(equal(4))
+        expect(virtualCurrenciesAfterPurchase["TEST"]?.code).to(equal("TEST"))
+        expect(virtualCurrenciesAfterPurchase["TEST"]?.name).to(equal("Test Currency"))
+        expect(virtualCurrenciesAfterPurchase["TEST"]?.serverDescription).to(equal("This is a test currency"))
+        assertVCCodeHasNoBalance(virtualCurrenciesAfterPurchase, vcCode: "TEST2")
+
+        // Ensure that this purchase didn't grant VCs to other subscribers
+        _ = try await self.purchases.logIn(userID2)
+        let virtualCurrenciesForOtherUser = try await self.purchases.virtualCurrencies()
+
+        assertAllVirtualCurrenciesHaveZeroBalances(virtualCurrenciesForOtherUser)
+    }
+
+    func testDoesntGrantVCForProductWithoutVCGrant() async throws {
+        let userID1 = "vc_user_\(UUID().uuidString)"
+        _ = try await self.purchases.logIn(userID1)
+
+        try self.purchases.invalidateVirtualCurrenciesCache()
+        let virtualCurrenciesBeforePurchase = try await self.purchases.virtualCurrencies()
+        assertAllVirtualCurrenciesHaveZeroBalances(virtualCurrenciesBeforePurchase)
+
+        let product = try await StoreKit.Product.products(for: [Self.weeklyWith3DayTrial]).first!
+        _ = try await self.purchases.purchase(product: StoreProduct(sk2Product: product))
+
+        try self.purchases.invalidateVirtualCurrenciesCache()
+
+        let virtualCurrenciesAfterPurchase = try await self.purchases.virtualCurrencies()
+        assertAllVirtualCurrenciesHaveZeroBalances(virtualCurrenciesAfterPurchase)
+    }
+
+    // MARK: - Assertion Helpers
+    private func assertAllVirtualCurrenciesHaveZeroBalances(_ virtualCurrencies: VirtualCurrencies) {
+        assertVCCodeHasNoBalance(virtualCurrencies, vcCode: "TEST")
+        assertVCCodeHasNoBalance(virtualCurrencies, vcCode: "TEST2")
+    }
+
+    private func assertVCCodeHasNoBalance(
+        _ virtualCurrencies: VirtualCurrencies,
+        vcCode: String
+    ) {
+        // swiftlint:disable:next todo
+        // TODO: We might need to check for this to be 0 depending on the outcome from a discussion with the team
+        expect(virtualCurrencies[vcCode]?.balance).to(beNil())
```

Comment:
Currently, for performance reasons, the backend sends down an empty dictionary for users who either:
- Have made no purchases
- Have received no virtual currency grants
We're currently discussing changing this so that the backend always returns all virtual currencies, but until then, this assertion just checks that the balance is nil. If the change is made in the backend, this would instead check that the balance is 0.